modules/distribution/keeper/querier.go

```diff
@@ -0,0 +1,242 @@
+	}
+	ddi := k.GetDelegationDistInfo(ctx, params.DelegatorAddress, params.ValidatorAddress)
+	res, errRes := codec.MarshalJSONIndent(k.cdc, ddi)
+	if errRes != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr("could not marshal result to JSON", errRes.Error()))
+	}
+	return res, nil
+}
+
+func queryAllDelegationDistInfo(ctx sdk.Context, _ []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params QueryDelegatorParams
+	err := k.cdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	// cache-wrap context as to not persist state changes during querying
+	ctx, _ = ctx.CacheContext()
+	var distInfos []types.DelegationDistInfo
+	ddiIter := func(_ int64, distInfo types.DelegationDistInfo) (stop bool) {
+		distInfos = append(distInfos, distInfo)
+		if err != nil {
+			panic(err)
+		}
+		return false
+	}
+	k.IterateDelegatorDistInfos(ctx, params.DelegatorAddress, ddiIter)
+	res, errRes := codec.MarshalJSONIndent(k.cdc, distInfos)
+	if errRes != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr("could not marshal result to JSON", errRes.Error()))
+	}
+	return res, nil
+}
+
+// params for query 'custom/distr/validator_dist_info'
+type QueryValidatorDistInfoParams struct {
+	ValidatorAddress sdk.ValAddress `json:"validator_address"`
+}
+
+// creates a new instance of QueryValidatorDistInfoParams
+func NewQueryValidatorDistInfoParams(validatorAddr sdk.ValAddress) QueryValidatorDistInfoParams {
+	return QueryValidatorDistInfoParams{
+		ValidatorAddress: validatorAddr,
+	}
+}
+
+func queryValidatorDistInfo(ctx sdk.Context, _ []string, req abci.RequestQuery, k Keeper) ([]byte, sdk.Error) {
+	var params QueryValidatorDistInfoParams
+	err := k.cdc.UnmarshalJSON(req.Data, &params)
+	if err != nil {
+		return nil, sdk.ErrUnknownRequest(sdk.AppendMsgToErr("incorrectly formatted request data", err.Error()))
+	}
+
+	// cache-wrap context as to not persist state changes during querying
+	ctx, _ = ctx.CacheContext()
+	if !k.HasValidatorDistInfo(ctx, params.ValidatorAddress) {
+		return []byte{}, types.ErrNoValidatorDistInfo(types.DefaultCodespace)
+	}
+	vdi := k.GetValidatorDistInfo(ctx, params.ValidatorAddress)
+	res, errRes := codec.MarshalJSONIndent(k.cdc, vdi)
+	if errRes != nil {
+		return nil, sdk.ErrInternal(sdk.AppendMsgToErr("could not marshal result to JSON", errRes.Error()))
+	}
+	return res, nil
+}
+
+// params for query 'custom/distr/rewards'
+type QueryRewardsParams struct {
+	Address sdk.AccAddress `json:"address"`
+}
+
+// creates a new instance of QueryRewardsParams
+func NewQueryRewardsParams(address sdk.AccAddress) QueryRewardsParams {
+	return QueryRewardsParams{
+		Address: address,
+	}
+}
+
+type Rewards struct {
+	Total       sdk.Coins           `json:"total"`
+	Delegations []DelegationsReward `json:"delegations"`
+	Commission  sdk.Coins           `json:"commission"`
+}
```

Comment:
Due to decimal truncation, `total` may not equal to the sum of delegations reward and commission. Is this acceptable?

Reply:
I think no problem, any good suggestions?